distributed/scheduler.py

```diff
@@ -4910,7 +4991,20 @@ def validate_state(self, allow_overlap: bool = False) -> None:
                 assert not ws.occupancy
                 if ws.status == Status.running:
                     assert ws.address in self.idle
+            assert not (set(ws.needs_what) & set(ws.has_what))
+            actual_needs_what: defaultdict[TaskState, int] = defaultdict(int)
+            for ts in ws.processing:
+                for tss in ts.dependencies:
+                    if tss not in ws.has_what:
+                        actual_needs_what[tss] += 1
+            assert actual_needs_what == ws.needs_what
             assert (ws.status == Status.running) == (ws in self.running)
+            for name, count in ws.task_groups_count.items():
+                task_group_counts[name] += count
+
+        assert task_group_counts.keys() == self.task_groups_count_global.keys()
+        for name, global_count in self.task_groups_count_global.items():
+            assert task_group_counts[name] == global_count, name
```

Comment:
pytest only perfoms assertion rewriting for test modules to avoid altering bytecode of the SUT, so this just spits out a plain `AssertionError` per default. (https://docs.pytest.org/en/latest/how-to/writing_plugins.html?highlight=pytest.register_assert_rewrite#assertion-rewriting)